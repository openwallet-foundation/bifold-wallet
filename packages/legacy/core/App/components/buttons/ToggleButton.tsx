import React, { useState, useEffect } from 'react'
import { Pressable, Animated } from 'react-native'
import Icon from 'react-native-vector-icons/MaterialIcons'
import { useTheme } from '../../contexts/theme'
import { useTranslation } from 'react-i18next'

interface ToggleButtonProps {
  isEnabled: boolean
  isAvailable: boolean
  toggleAction: () => void
  testID?: string
  enabledIcon?: string
  disabledIcon?: string
  disabled?: boolean
}

const ToggleButton: React.FC<ToggleButtonProps> = ({
  isEnabled,
  isAvailable,
  toggleAction,
  testID,
  enabledIcon = 'check',
  disabledIcon = 'close',
  disabled = false,
}) => {
  const { ColorPallet } = useTheme()
  const [toggleAnim] = useState(new Animated.Value(isEnabled ? 1 : 0))
  const { t } = useTranslation()

  useEffect(() => {
    Animated.timing(toggleAnim, {
      toValue: isEnabled ? 1 : 0,
      duration: 200,
      useNativeDriver: true,
    }).start()
  }, [isEnabled, toggleAnim])

  const backgroundColor = toggleAnim.interpolate({
    inputRange: [0, 1],
    outputRange: [ColorPallet.grayscale.lightGrey, ColorPallet.brand.primary],
  })

  const translateX = toggleAnim.interpolate({
    inputRange: [0, 1],
    outputRange: [2, 25],
  })

  return (
    <Pressable
      accessible
      testID={testID}
      accessibilityLabel={isEnabled ? t('Biometry.On') : t('Biometry.Off')}
      accessibilityRole="switch"
      accessibilityState={{
        checked: isEnabled,
      }}
      onPress={isAvailable && !disabled ? toggleAction : undefined} // Prevent onPress if not available or disabled
      disabled={!isAvailable || disabled}
    >
      <Animated.View
        style={{
          width: 55,
          height: 30,
          borderRadius: 25,
          backgroundColor,
          padding: 3,
          justifyContent: 'center',
          opacity: disabled ? 0.5 : 1, // Visual feedback for disabled state
        }}
      >
        <Animated.View
          style={{
            transform: [{ translateX }],
            width: 25,
            height: 25,
            borderRadius: 20,
            backgroundColor: '#FFFFFF',
            justifyContent: 'center',
            alignItems: 'center',
          }}
        >
          <Icon
            name={isEnabled ? enabledIcon : disabledIcon}
            size={15}
            color={isEnabled ? ColorPallet.brand.primary : ColorPallet.grayscale.mediumGrey}
          />
        </Animated.View>
      </Animated.View>
    </Pressable>
  )
}

export default ToggleButton
