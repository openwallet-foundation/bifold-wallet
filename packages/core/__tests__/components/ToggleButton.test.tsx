import { render, fireEvent } from '@testing-library/react-native'
import React from 'react'
import ToggleButton from '../../src/components/buttons/ToggleButton'

describe('ToggleButton Component', () => {
  const toggleActionMock = jest.fn()

  beforeEach(() => {
    jest.clearAllMocks()
  })

  test('renders correctly when enabled', () => {
    const tree = render(
      <ToggleButton testID="toggle-button" isEnabled={true} isAvailable={true} toggleAction={toggleActionMock} />
    )
    expect(tree).toMatchSnapshot()
  })

  test('renders correctly when disabled', () => {
    const tree = render(
      <ToggleButton testID="toggle-button" isEnabled={false} isAvailable={true} toggleAction={toggleActionMock} />
    )
    expect(tree).toMatchSnapshot()
  })

  test('renders correctly when not available', () => {
    const tree = render(
      <ToggleButton testID="toggle-button" isEnabled={false} isAvailable={false} toggleAction={toggleActionMock} />
    )
    expect(tree).toMatchSnapshot()
  })

  test('calls toggleAction function when pressed and available', () => {
    const { getByTestId } = render(
      <ToggleButton testID="toggle-button" isEnabled={false} isAvailable={true} toggleAction={toggleActionMock} />
    )
    const toggle = getByTestId('toggle-button')
    fireEvent.press(toggle)
    expect(toggleActionMock).toHaveBeenCalled()
  })

  test('does not call toggleAction when unavailable', () => {
    const { getByTestId } = render(
      <ToggleButton testID="toggle-button" isEnabled={false} isAvailable={false} toggleAction={toggleActionMock} />
    )
    const toggle = getByTestId('toggle-button')
    fireEvent.press(toggle)
    expect(toggleActionMock).not.toHaveBeenCalled()
  })

  test('does not call toggleAction when disabled', () => {
    const { getByTestId } = render(
      <ToggleButton
        testID="toggle-button"
        isEnabled={false}
        isAvailable={true}
        toggleAction={toggleActionMock}
        disabled={true}
      />
    )
    const toggle = getByTestId('toggle-button')
    fireEvent.press(toggle)
    expect(toggleActionMock).not.toHaveBeenCalled()
  })

  test('renders custom icons correctly', () => {
    const { getByTestId } = render(
      <ToggleButton
        testID="toggle-button"
        isEnabled={true}
        isAvailable={true}
        toggleAction={toggleActionMock}
        enabledIcon="thumb-up"
        disabledIcon="thumb-down"
      />
    )
    const toggle = getByTestId('toggle-button')
    expect(toggle).toBeTruthy()
  })
})
