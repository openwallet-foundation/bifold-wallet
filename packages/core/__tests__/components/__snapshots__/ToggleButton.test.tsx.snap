// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToggleButton Component renders correctly when disabled 1`] = `
<View
  accessibilityLabel="Biometry.Off"
  accessibilityRole="switch"
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": false,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="toggle-button"
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "rgba(211, 211, 211, 1)",
        "borderRadius": 25,
        "height": 30,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 3,
        "width": 55,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 20,
          "height": 25,
          "justifyContent": "center",
          "transform": Array [
            Object {
              "translateX": 1,
            },
          ],
          "width": 25,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#606060",
              "fontSize": 15,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`ToggleButton Component renders correctly when enabled 1`] = `
<View
  accessibilityLabel="Biometry.On"
  accessibilityRole="switch"
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": true,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="toggle-button"
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "rgba(66, 128, 62, 1)",
        "borderRadius": 25,
        "height": 30,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 3,
        "width": 55,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 20,
          "height": 25,
          "justifyContent": "center",
          "transform": Array [
            Object {
              "translateX": 24,
            },
          ],
          "width": 25,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#42803E",
              "fontSize": 15,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;

exports[`ToggleButton Component renders correctly when not available 1`] = `
<View
  accessibilityLabel="Biometry.Off"
  accessibilityRole="switch"
  accessibilityState={
    Object {
      "busy": undefined,
      "checked": false,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    Object {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="toggle-button"
>
  <View
    collapsable={false}
    style={
      Object {
        "backgroundColor": "rgba(211, 211, 211, 1)",
        "borderRadius": 25,
        "height": 30,
        "justifyContent": "center",
        "opacity": 1,
        "padding": 3,
        "width": 55,
      }
    }
  >
    <View
      collapsable={false}
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#FFFFFF",
          "borderRadius": 20,
          "height": 25,
          "justifyContent": "center",
          "transform": Array [
            Object {
              "translateX": 1,
            },
          ],
          "width": 25,
        }
      }
    >
      <Text
        allowFontScaling={false}
        selectable={false}
        style={
          Array [
            Object {
              "color": "#606060",
              "fontSize": 15,
            },
            undefined,
            Object {
              "fontFamily": "Material Icons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        
      </Text>
    </View>
  </View>
</View>
`;
